securebanking-openbanking-uk-gateway-conversion-filter/src/main/java/com/forgerock/securebanking/uk/gateway/conversion/filter/IntentConverterFilter.java

```diff
@@ -0,0 +1,110 @@
+import org.forgerock.http.protocol.Status;
+import org.forgerock.openig.heap.GenericHeaplet;
+import org.forgerock.services.context.Context;
+import org.forgerock.util.promise.NeverThrowsException;
+import org.forgerock.util.promise.Promise;
+import org.forgerock.util.promise.PromiseImpl;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.Set;
+
+import static org.forgerock.http.protocol.Response.newResponsePromise;
+import static org.forgerock.json.JsonValueFunctions.enumConstant;
+
+/**
+ * Filter to convert IDM json intent objects to OB data model objects.
+ *
+ * This filter must have received {@code intentType} as required to identify the intent type {@link IntentType#toString()} to instance the converter,
+ * the {@code intentContent} as optional of string representation of a json intent to be converter to OB data model object,
+ * if 'intentContent' isn't provided the filter will get the intent content in string format from the entity request.
+ * Failures from the `Converter instance` is either a {@link RuntimeException} that will catch to build a {@link ResponseException}.
+ *
+ * Configuration options:
+ *
+ * <pre>
+ * {@code {
+ *      "type": "ConversionToOBObjectFilter",
+ *      "config": {
+ *         "intentType"     IntentType#toString()    [REQUIRED - (ItIdentifies the consent type by enum name (ACCOUNT_ACCESS_CONSENT, PAYMENT_INTERNATIONAL_CONSENT... ]
+ *         "intentContent"  string                   [OPTIONAL - String representation of json intent object to be converted to OB Object]
+ *      }
+ *  }
+ *  }
+ * </pre>
+ */
+public class IntentConverterFilter implements Filter {
+
+    private static final Logger logger = LoggerFactory.getLogger(IntentConverterFilter.class);
+
+    private final IntentType intentType;
+    private final String intentContent;
+
+    public IntentConverterFilter(final IntentType intentType, final String stringObject) {
+        this.intentType = intentType;
+        this.intentContent = stringObject;
+    }
+
+    @Override
+    public Promise<Response, NeverThrowsException> filter(Context context, Request request, Handler next) {
+        // To be able to compile with the generics type erasure, we have to get the help of another PromiseImpl.
+        PromiseImpl<Set<String>, ResponseException> promise = PromiseImpl.create();
+        String entity = intentContent;
+        try {
+            if (entity == null) {
+                entity = request.getEntity().getString();
+            }
+            Object converted = ConverterFactory.getConverter(intentType).convertFromJsonString(entity);
+            request.setEntity(converted);
+            // add the entity to the response
+            return next.handle(context, request)
+                    .thenOnResult(response -> {
+                        response.setEntity(converted);
+                        response.setStatus(Status.OK);
+                    });
```

Comment:
I think the response conversion needs some thought, the current impl will send the converted intent attached to the request (or from the config) in the response. However I would assume that the response may be a different object from the request? 

We may only want to support converting the request or response payload (but not both), as the response may already be a valid OB object. For instance, when calling into the back office API to calculate charges for a payment, the request object will start off as an IDM json and then be converted to OB obj to send to the back office, I assume the back office then replies with an OB object (so no conversion on the response path is required).

Please correct me if I have misunderstood the requirements / how the impl works...

Reply:
Cases covered in this solution:
- A case when the TPP request the consent pushed previously (to check the status for example, means GET the Consent), where will not necessary call the RS backend to build the response, the TPP will retrieve a consistent OB data model object from the response.
- A case where we need to build the OB data model object to call RS, in that case we can define in the API request body and in the `ResponseEntity` as an OB data model object instead of a String as is defined currently:
```java
// currently
ResponseEntity calculateElements(
            @ApiParam(value = "Create an Account Access Consents", required = true)
            @Valid
            @RequestBody String body,
.....
// Future (calculate elements and update the ob object with calculated elements to return it to IG and update the intent on IDM)
ResponseEntity<OBReadConsentResponse1> calculateElements(
            @ApiParam(value = "Create an Account Access Consents", required = true)
            @Valid
            @RequestBody OBReadConsentResponse1 body,
```
this will generate a more consistent swagger documentation for the customer if they want to implement his own RS.
It means also that we need to modify the RS backend properly.

Reply:
The use cases make sense, I am just concerned with the logic. 

I am not familiar with IG filters so perhaps I am misunderstanding how this new filter will be called, when this filter is added to an existing chain then I should be able to understand the flow better.

My concern is with lines 86/87 in the snippet, when the filter is called as part of a chain, how do we know that we always want to set the response entity and status to the converted request entity and 200 response? Can't filters further into the chain potentially have mutated (or produced an entirely different response object) or have raised an error?

For example, given a filter chain:
- request -> FilterA -> IntentConverterFilter -> FilterB -> SomeHandler
- SomeHandler -> FilterB -> IntentConverterFilter -> FilterA-> response

Then if FilterB or SomeHandler produce/edit the response entity (and/or status code), then the IntentConverterFilter will clobber these values.

Therefore we may want to have some configuration around when the filter is applied (either through different classes or some flag), so that we can control if the converter logic is applied to request only, responses only or both.

Reply:
New field `resultTo` introduced to indicate where will be set the conversion result object (request, response or both)
```json
{
    "name": "IntentConverterFilter-accessAccountConsent"
     "type": "IntentConverterFilter",
    "config": {
        "intentType": "ACCOUNT_ACCESS_CONSENT",
        "payloadFrom":"REQUEST",
        "resultTo": ["RESPONSE"]
    }
}
```